src/NodaTime/Calendars/WeekYearRule.cs

```diff
@@ -0,0 +1,199 @@
+    /// A rule determining how "week years" are arranged, including the weeks within the week year.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// Dates are usually identified within a calendar system by a calendar year, a month within that
+    /// calendar year, and a day within that month. For example, the date of birth of Ada Lovelace can be identified
+    /// within the Gregorian calendar system as the year 1815, the month December (12), and the day 10. However,
+    /// dates can also be identified (again within a calendar system) by week-year, week and day-of-week. How
+    /// that identification occurs depends on which rule you use - but again as an example, within the Gregorian
+    /// calendar system, using the ISO-8601 week year rule, the date of the Unix epoch is week-year 1815,
+    /// week 49, day-of-week Sunday.
+    /// </para> 
+    /// <para>
+    /// The calendar year of a date and the week-year of a date are the same in most rules for most dates, but aren't
+    /// always. When they differ, it is usually because a day near the start of the calendar year is deemed to belong
+    /// to the last week of the previous week-year - or conversely because a day near the end of the calendar year is
+    /// deemed to belong to the first week of the following week-year. Some rules may be more radical -
+    /// a UK tax year rule could number weeks from April 6th onwards, such that any date earlier than that in the calendar
+    /// year would belong to the previous week-year.
+    /// </para>
+    /// <para>
+    /// The mapping of dates into week-year, week and day-of-week is always relative to a specific calendar system.
+    /// For example, years in the Hebrew calendar system vary very significantly in length due to leap months, and this
+    /// is reflected in the number of weeks within the week-years - as low as 50, and as high as 55.
+    /// </para>
+    /// <para>
+    /// This class allows conversions between the two schemes of identifying dates: <see cref="GetWeekYear(LocalDate)"/>
+    /// and <see cref="GetWeekOfWeekYear(LocalDate)"/> allow the week-year and week to be obtained, and
+    /// <see cref="CreateLocalDate(int, int, IsoDayOfWeek, CalendarSystem)"/> allows the reverse mapping. Note that
+    /// the calendar system does not need to be specified in the former methods as a <see cref="LocalDate"/> already
+    /// contains calendar information, and there is no method to obtain the day-of-week as that is not affected by the
+    /// week year rule being used.
+    /// </para>
+    /// <para>Specific rules are obtained using the static factory methods within this class, or with the
+    /// <see cref="Iso"/> property.</para>
+    /// <para>
+    /// All implementations within Noda Time are immutable, and it is advised that any external implementations
+    /// should be immutable too.
+    /// </para>
+    /// </remarks>
+    [Immutable]
+    public abstract class WeekYearRule
```

Comment:
This is coupling the user-facing API with our implementation, which is the same thing we did with `DateTimeZone`.

Would probably prefer an interface here (and extension method for `GetWeeksInWeekYear(int)`?), and then if we later decide it's useful to provide one or more abstract base classes for external implementations to make use of, we can do so without breaking existing implementations.